Error from function call with incorrect return type outside braces in unit valued match suggests semicolon to be added; suggestion is incorrect

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

This small snippet

```rust
fn func() -> u8 {
    0
}

fn main() {
    match () {
        () => func()
    }
}
```

suggests adding a semicolon after the function call to rectify the mismatched types.

```
error[E0308]: mismatched types
 --> src/main.rs:7:15
  |
5 | fn main() {
  |           - expected `()` because of default return type
6 |     match () {
7 |         () => func()
  |               ^^^^^^- help: consider using a semicolon here: `;`
  |               |
  |               expected `()`, found `u8`

error: aborting due to previous error
```

However, this is also a syntax error (below), and the correct suggestion would be to wrap the function call in braces.

```
error: `match` arm body without braces
 --> src/main.rs:7:15
  |
7 |         () => func();
  |            -- ^^^^^^- help: use a comma to end a `match` arm expression: `,`
  |            |  |
  |            |  this statement is not surrounded by a body
  |            while parsing the `match` arm starting here

error: aborting due to previous error
```

### Meta
Using rustc nightly `2021-04-04 c755ee4ce8cae6ea977d` (playground)

--- a/compiler/rustc_typeck/src/check/coercion.rs
+++ b/compiler/rustc_typeck/src/check/coercion.rs
@@ -1440,9 +1440,8 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
         // as prior return coercions would not be relevant (#57664).
         let parent_id = fcx.tcx.hir().get_parent_node(id);
         let fn_decl = if let Some((expr, blk_id)) = expression {
-            pointing_at_return_type = fcx.suggest_mismatched_types_on_tail(
-                &mut err, expr, expected, found, cause.span, blk_id,
-            );
+            pointing_at_return_type =
+                fcx.suggest_mismatched_types_on_tail(&mut err, expr, expected, found, blk_id);
             let parent = fcx.tcx.hir().get(parent_id);
             if let (Some(cond_expr), true, false) = (
                 fcx.tcx.hir().get_if_cause(expr.hir_id),

--- a/compiler/rustc_typeck/src/check/expr.rs
+++ b/compiler/rustc_typeck/src/check/expr.rs
@@ -603,7 +603,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         &cause,
                         &mut |mut err| {
                             self.suggest_mismatched_types_on_tail(
-                                &mut err, expr, ty, e_ty, cause.span, target_id,
+                                &mut err, expr, ty, e_ty, target_id,
                             );
                             if let Some(val) = ty_kind_suggestion(ty) {
                                 let label = destination

--- a/compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs
@@ -41,15 +41,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expr: &'tcx hir::Expr<'tcx>,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
-        cause_span: Span,
         blk_id: hir::HirId,
     ) -> bool {
         let expr = expr.peel_drop_temps();
         // If the expression is from an external macro, then do not suggest
         // adding a semicolon, because there's nowhere to put it.
         // See issue #81943.
-        if expr.can_have_side_effects() && !in_external_macro(self.tcx.sess, cause_span) {
-            self.suggest_missing_semicolon(err, expr, expected, cause_span);
+        if expr.can_have_side_effects() && !in_external_macro(self.tcx.sess, expr.span) {
+            self.suggest_missing_semicolon(err, expr, expected);
         }
         let mut pointing_at_return_type = false;
         if let Some((fn_decl, can_suggest)) = self.get_fn_decl(blk_id) {
@@ -388,7 +387,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         err: &mut DiagnosticBuilder<'_>,
         expression: &'tcx hir::Expr<'tcx>,
         expected: Ty<'tcx>,
-        cause_span: Span,
     ) {
         if expected.is_unit() {
             // `BlockTailExpression` only relevant if the tail expr would be
@@ -403,7 +401,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     if expression.can_have_side_effects() =>
                 {
                     err.span_suggestion(
-                        cause_span.shrink_to_hi(),
+                        expression.span.shrink_to_hi(),
                         "consider using a semicolon here",
                         ";".to_string(),
                         Applicability::MachineApplicable,

--- a/src/test/ui/suggestions/issue-83892.fixed
+++ b/src/test/ui/suggestions/issue-83892.fixed
@@ -1,0 +1,11 @@
+// run-rustfix
+
+fn func() -> u8 {
+    0
+}
+
+fn main() {
+    match () {
+        () => func() //~ ERROR mismatched types
+    };
+}

--- a/src/test/ui/suggestions/issue-83892.rs
+++ b/src/test/ui/suggestions/issue-83892.rs
@@ -1,0 +1,11 @@
+// run-rustfix
+
+fn func() -> u8 {
+    0
+}
+
+fn main() {
+    match () {
+        () => func() //~ ERROR mismatched types
+    }
+}

--- a/src/test/ui/suggestions/issue-83892.stderr
+++ b/src/test/ui/suggestions/issue-83892.stderr
@@ -1,0 +1,14 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-83892.rs:9:15
+   |
+LL | fn main() {
+   |           - expected `()` because of default return type
+LL |     match () {
+LL |         () => func()
+   |               ^^^^^^ expected `()`, found `u8`
+LL |     }
+   |      - help: consider using a semicolon here: `;`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

